### PR TITLE
Refactor how variables are added to the model object

### PR DESCRIPTION
**Pull request by tomicapretto**
Currently, `Model._add()` is a long method that does a lot of stuff and I think it that a first sight it not clear what is going on. In this PR I'm adding a variety of changes to make `Model._add()` clearer as well as solving #275, #271 and (maybe) #224

A list of (planned) changes

* [x] Split `Model._add()` into `Model._add_fixed()` and `Model._add_random()`
* [x] Added `get_bernoulli_data()` to `utils.py`. This function is called to handle the response variable when `family='bernoulli'`.
* [x] Add more tests. For example, check Bambi works when no response level is explicitly passed and `family='bernoulli'`.
* [x] Added `ResponseTerm` class, see comments.
* [x] Add tests for new plot labels

I deleted this point 

* [Use `Term` and `RandomTerm` consistently (see https://github.com/bambinos/bambi/issues/275#issuecomment-734498678)

Because (as mentioned below) I think it should be part of another refactor.

### CODE REVIEW

**Review by twiecki — APPROVED** [review]

**Comment by twiecki** [comment]
@tomicapretto This looks like a great contirbution and useful refactoring.

Note that this project requires black:
```
hecking code style with black....

+python -m black bambi --check

would reformat /home/travis/build/bambinos/bambi/bambi/tests/test_built_models.py

Oh no! 💥 💔 💥
```

**Comment by tomicapretto** [comment]
I've just added a new class `ResponseTerm` that inherits from `Term` that is only used for the response variable. I think the information about the level that is being modelled should be saved in that object and not inferred as it was done before, i. e.

```python
        # For bernoulli models with n_trials = 1 (most common use case),
        # tell user which event is being modeled
        if self.family.name == "bernoulli" and np.max(self.y.data) < 1.01:
            event = next(i for i, x in enumerate(self.y.data.flatten()) if x > 0.99)
            _log.info(
                "Modeling the probability that %s==%s",
                self.y.name,
                str(self.clean_data[self.y.name].iloc[event]),
            )
```

Now it is 
```python
        # Tell user which event is being modeled
        if self.family.name == "bernoulli":
            _log.info(
                "Modeling the probability that %s==%s",
                self.y.name,
                str(self.y.success_event),
            )
```

Also, currently all response variables have `categorical` equal to `False`. Now, if we are in a model where `family='bernoulli'`, the response term has `True` for `categorical`. Finally, Bambi raises an early error when try to use index notation with other family than 'bernoulli'.


I haven't discussed this changes with anyone before this, so please let me know if anything in this PR does not make sense and we discuss about the approach to follow.

**Comment by tomicapretto** [comment]
Some updates: 

1. I could get better labels for categorical variables that are used as fixed predictors as well as random terms.
Have a look at the following:

```python
import arviz as az
import bambi as bmb
import numpy as np
import pandas as pd

np.random.seed(1234)
az.style.use('arviz-darkgrid')

data = pd.DataFrame({
    'y' : np.random.normal(size=50),
    'x' : np.random.normal(size=50),
    'z' : ['Group 1'] * 10 + ['Group 2'] * 10 + ['Group 3'] * 10 + ['Group 1'] * 10 + ['Group 2'] * 10,
    'time': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10] * 5,
    'subject': ['aaa'] * 10 + ['bbb'] * 10 + ['ccc'] * 10 + ['ddd'] * 10 + ['eee'] * 10
})
model = bmb.Model(data)
fitted = model.fit('y ~ x + z + time', random=['time|subject'])

model.plot_priors();
```
priors

```python
az.plot_trace(fitted);
```
posteriors

2. While it helps with the plots, it does not help with the summary tables. It looks like some information is lost in the middle. I haven't figured out how to fix this yet.

```python
az.summary(fitted)
```

image

3. I still need to add tests. 

Finally, while what I've done seems to work,  there may be better ways. Please let me know if anything, either efficiency or style,  could be improved.

File: bambi/models.py [hunk 1]
```diff
@@ -615,14 +500,138 @@ def _add_y(self, variable, prior=None, family="gaussian", link=None, *args, **kw
...
+        categorical = False
+
+        if event is not None:
+            if self.family.name != "bernoulli":
+                raise ValueError("Index notation only available for 'bernoulli' family")
+            # pass in new Y data that has 1 if y=event and 0 otherwise
+            success_event = event.group(1)
+            categorical = True
+            data = vector[:, vector.design_info.column_names.index(success_event)]
+            data = pd.DataFrame({event.group(3): data})
```
Comment (ravinkumar) [inline bambi/models.py:519]: What is group 3? Can a comment be added here for clarification?

**Review by ravinkumar — APPROVED** [review]

File: bambi/models.py (hunk 1 above)
Comment (tomicapretto) [reply bambi/models.py:519]: It is already written in `_add_fixed()` method. 

```python
    def _add_fixed(self, fixed, data, family, link, priors):
        if "~" in fixed:
            # check to see if formula is using the 'y[event] ~ x' syntax.
            # If so, chop it into groups:
            # 1 = 'y[event]', 2 = 'y', 3 = 'event', 4 = 'x'
            # If this syntax is not being used, event = None
            event = re.match(r"^((\S+)\[(\S+)\])\s*~(.*)$", fixed)
```
But I think an reminder there will be helpful. Thanks!!

Comment (ravinkumar) [reply bambi/models.py:519]: Oh I see. my apologies. Treat this comment as optional then

**Comment by aloctavodia** [comment]
What you are seeing with the summary function is related to arviz-devs/arviz#1201

File: bambi/utils.py [hunk 2]
```diff
@@ -7,3 +13,36 @@ def listify(obj):
...
+    else:
+        raise ValueError("Response variable is of the wrong type.")
+    return data, success
+
+
+def extract_label(string, kind="fixed"):
+    """Receives a level as returned by patsy and returns a cleaned version of it.
+    Fixed and random effects are handled differently because their representations differ.
+    """
+    assert kind in ["fixed", "random"], "kind must be 'fixed' or 'random'"
```
Comment (aloctavodia) [inline bambi/utils.py:38]: I think we can safely remove this assertion. and move "kind must be 'fixed' or 'random'" to the docstring

File: bambi/tests/test_built_models.py [hunk 3]
```diff
@@ -421,6 +421,36 @@ def dicts_close(a, b):
     assert all([dicts_close(priors0[x], priors1[x]) for x in priors0.keys()])
 
 
+def test_logistic_regression_empty_index():
+    np.random.seed(303456)
```
Comment (aloctavodia) [inline bambi/tests/test_built_models.py:425]: We can remove the random seeds from these tests

File: bambi/models.py [hunk 4]
```diff
@@ -1020,6 +1037,25 @@ def __init__(self, name, data, categorical=False, prior=None, constant=None):
...
+
+    def clean_levels(self):
+        if self.random:
+            self.levels = [extract_label(level, "random") for level in self.levels]
+        else:
+            self.levels = [extract_label(level, "fixed") for level in self.levels]
+
+
+class ResponseTerm(Term):
+    def __init__(self, name, data, categorical=False, prior=None, success_event=None):
```
Comment (aloctavodia) [inline bambi/models.py:1050]: add docstring

File: bambi/models.py [hunk 5]
```diff
@@ -615,14 +500,139 @@ def _add_y(self, variable, prior=None, family="gaussian", link=None, *args, **kw
...
+
+            if EvalFactor(_name) in factor_infos:
+                categorical = factor_infos[EvalFactor(_name)].type == "categorical"
+            else:
+                categorical = False
+            prior = priors.pop(_name, priors.get("fixed", None))
+            self.terms[_name] = Term(_name, term_data, categorical=categorical, prior=prior)
+
+    def _add_random(self, random, data, priors):
+        for random_effect in random:  # pylint: disable=too-many-nested-blocks
```
Comment (aloctavodia) [inline bambi/models.py:559]: I think we no longer need to disable this

**Comment by tomicapretto** [comment]
While everything seems to work (all tests have passed) I think this issue is not completely solved. Some reasons

1. Now, Term-like objects have an attribute `cleaned_levels` and `levels`. The first is used  when creating the PyMC model and the latters is used a lot internally. That's why I had to leave `levels`, because many things depended on how it worked. I think this will be completely solved the day we make a greater refactor.
2. Interaction terms have more or less redundant labels. They appear like "var_a:var_b\n var_a[level_1]:var_b[level_2]. 
3. The tabulates produced with `az.summary()` still has numeric index. I know this is related to ArviZ, but in the end, the problem is not completely solved.

My next step is to add more tests related to all these functionalities and I think we will be able to merge. In the (hopefully, near) future I would like to refactor more things so we don't break much things when we want to introduce other changes in the future.

Any suggestions?

**Comment by aloctavodia** [comment]
option 1, we merge once test pass and prepare a new release.
option 2, we put this on hold, make a release and then work on a more profound refactor.

I think I prefer option 1 as this PR seems to be an improvement, even if it only partially solve the problem.

**Comment by twiecki** [comment]
@aloctavodia I agree, I think this is a solid piece of work that improves the status quo. Delaying it might just cause us to lose momentum on this.

**Comment by tomicapretto** [comment]
I think this is done now. I've been trying to refactor how we add random terms a little but I realized it is more appropriate for a deeper refactor.

**Review by twiecki — APPROVED** [review]

**Comment by twiecki** [comment]
Thanks @tomicapretto, this looks great to me.

**Pull request merged**